src/js/node/zlib.ts

```diff
@@ -1,4610 +1,453 @@
+    }
   }
-  if (typeof callback !== "function") throw new TypeError("BrotliDecoder callback is not callable");
-  const decoder = createBrotliDecoder(opts, {}, callback);
-  decoder.decode(buffer, undefined, true);
+  const [, , private_constructor] = methods[method];
+  this[kHandle] = private_constructor(options, {}, null, method);
+  stream.Transform.$call(this, options);
 }
+Base.prototype = Object.create(stream.Transform.prototype);
+ObjectDefineProperty(Base.prototype, "_handle", {
+  get: function () {
+    return this[kHandle];
+  },
+  set: function (newval) {
+    //noop
+  },
+});
+ObjectDefineProperty(Base.prototype, "bytesWritten", {
+  get: function () {
+    return this[kHandle].bytesWritten;
+  },
+});
+ObjectDefineProperty(Base.prototype, "bytesRead", {
+  get: function () {
+    return this[kHandle].bytesRead;
+  },
+});
+ObjectDefineProperty(Base.prototype, "_closed", {
+  get: function () {
+    return this[kHandle].closed;
+  },
+});
+Base.prototype.flush = function (kind, callback) {
+  if (typeof kind === "function" || (kind === undefined && !callback)) {
+    callback = kind;
+    kind = 3;
+  }
+  if (this.writableFinished) {
+    if (callback) process.nextTick(callback);
+  } else if (this.writableEnded) {
+    if (callback) this.once("end", callback);
+  } else {
+    this.write(kFlushBuffers[kind], "", callback);
+  }
+};
+Base.prototype.reset = function () {
+  assert(this[kHandle], "zlib binding closed");
+  return this[kHandle].reset();
+};
+Base.prototype.close = function (callback) {
+  if (callback) stream.finished(this, callback);
+  this.destroy();
+};
+Base.prototype._transform = function _transform(chunk, encoding, callback) {
+  try {
+    callback(undefined, this[kHandle].transformSync(chunk, encoding, false));
+  } catch (err) {
+    callback(err, undefined);
+  }
+};
+Base.prototype._flush = function _flush(callback) {
+  try {
+    callback(undefined, this[kHandle].transformSync("", undefined, true));
+  } catch (err) {
+    callback(err, undefined);
+  }
+};
+Base.prototype._final = function (callback) {
+  callback();
+};
+Base.prototype._processChunk = function (chunk, flushFlag, cb) {
+  // _processChunk() is left for backwards compatibility
+  if (typeof cb === "function") processChunk(this, chunk, flushFlag, cb);
+  else return processChunkSync(this, chunk, flushFlag);
+};
 
-function brotliCompressSync(buffer, opts) {
-  const encoder = createBrotliEncoder(opts, {}, null);
-  return encoder.encodeSync(buffer, undefined, true);
+function processChunkSync(self, chunk, flushFlag) {
+  return self[kHandle].transformSync(chunk, undefined, false, flushFlag);
 }
 
-function brotliDecompressSync(buffer, opts) {
-  const decoder = createBrotliDecoder(opts, {}, null);
-  return decoder.decodeSync(buffer, undefined, true);
+function processChunk(self, chunk, flushFlag, cb) {
+  if (self._closed) return process.nextTick(cb);
+  self[kHandle].transformSync(chunk, undefined, false, flushFlag);
```

Comment:
have you looked through node's source for this function? It doesn't look like they only call the callback when the handle is closed

https://github.com/nodejs/node/blob/2eff28fb7a93d3f672f80b582f664a7c701569fb/lib/zlib.js#L523-L542